zcash_transparent/src/builder.rs

```diff
@@ -378,4 +451,365 @@ impl Bundle<Unauthorized> {
+
+            // Apply the signature.
+            self.final_script_sigs[final_input_idx] = Some(construct_script_sig(
+                signature,
+                &self.authorization_inputs[final_input_idx].pubkey,
+            ));
+
+            Ok(self)
+        } else {
+            // This signature did not match any available unsigned inputs.
+            Err(Error::NoMatchingInputForSignature { sig_index: sig_idx })
+        }
+    }
+
+    /// Finalizes the signing process and attempts to build the `Bundle<Authorized>`.
+    ///
+    /// Returns an error if any signatures are missing.
+    pub fn finalize_signatures(self) -> Result<Bundle<Authorized>, Error> {
+        let mut fully_signed_scripts = Vec::with_capacity(self.final_script_sigs.len());
+        for script_opt in self.final_script_sigs.iter() {
+            match script_opt {
+                Some(script) => fully_signed_scripts.push(script.clone()),
+                None => {
+                    if !self.authorization_inputs.is_empty() {
+                        return Err(Error::NotAllInputsSigned);
+                    }
+                }
+            }
+        }
+
+        Ok(Bundle {
+            vin: self
+                .original_vin_unauthorized
+                .iter()
+                .zip(fully_signed_scripts)
+                .map(|(txin_unauth, sig_script)| TxIn {
+                    prevout: txin_unauth.prevout.clone(),
+                    script_sig: sig_script,
+                    sequence: txin_unauth.sequence,
+                })
+                .collect(),
+            vout: self.original_vout,
+            authorization: Authorized,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{Error, OutPoint, SignableInput, TransparentBuilder, TxOut};
+    use crate::address::TransparentAddress;
+    use ripemd::Ripemd160;
+    use secp256k1::{Message, Secp256k1, SecretKey};
+    use sha2::{Digest, Sha256};
+    use zcash_address::ZcashAddress;
+    use zcash_protocol::value::Zatoshis;
+
+    // Helper that takes a key to create distinct inputs.
+    fn new_p2pkh_spend_with_key(
+        key_bytes: [u8; 32],
+    ) -> (SecretKey, secp256k1::PublicKey, TxOut, OutPoint) {
+        let sk = SecretKey::from_slice(&key_bytes).expect("32 bytes is a valid secret key");
+        let secp = Secp256k1::new();
+        let pk = secp256k1::PublicKey::from_secret_key(&secp, &sk);
+
+        let pk_hash_generic_array = Ripemd160::digest(Sha256::digest(pk.serialize()));
+        let pk_hash_bytes: [u8; 20] = pk_hash_generic_array.into();
+        let taddr = TransparentAddress::PublicKeyHash(pk_hash_bytes);
+
+        let txout = TxOut {
+            value: Zatoshis::from_u64(10000).unwrap(),
+            script_pubkey: taddr.script(),
+        };
+
+        let txid = [0u8; 32]; // Dummy txid
+        let outpoint = OutPoint::new(txid, 0);
+
+        (sk, pk, txout, outpoint)
+    }
+
+    #[test]
+    fn happy_path_append_and_finalize_signatures() {
```

Comment:
```suggestion
    fn append_and_finalize_signatures() {
```
Style nit: it's sufficient to mark tests about failures by having "fails" in the name; it's not necessary to mark tests about successes.